### PR TITLE
feat: refactor for zksync abstraction

### PR DESCRIPTION
**Pull request by DanielSchiavini**
### What I did
This PR includes a few small changes to make zksync integration possible:
- Fix `prepare_calldata` for constructors
- Add `Env.create_deployer` method so it may be overridden
- Fix `sign_typed_data` in Google Colab (the address was getting lost)
- Change `sign_typed_data` to receive the full message instead of its components (the schema actually requires `primaryType`.
- Improve RPC error messages

### Cute Animal Picture
image

### CODE REVIEW

File: boa/contracts/vyper/compiler_utils.py [hunk 1]
```diff
@@ -104,43 +109,52 @@ def generate_bytecode_for_internal_fn(fn):
...
         @external
         @payable
         def __boa_debug__() {return_sig}:
             {debug_body}
     """
     )
-    return compile_vyper_function(wrapper_code, contract)
+
+
+def detect_statement_type(source_code, contract):
```
Comment (charles-cooper) [inline boa/contracts/vyper/compiler_utils.py:150]: probably should be renamed to `detect_expr_type`

File: boa/environment.py [hunk 2]
```diff
@@ -331,3 +336,23 @@ def time_travel(
 
         self.evm.patch.timestamp += seconds
         self.evm.patch.block_number += blocks
+
+    def create_deployer(
```
[the remark this answers is not in the document]
Comment (DanielSchiavini) [reply boa/environment.py]: Something like [this](https://github.com/vyperlang/titanoboa/pull/195/commits/debb3c86ba8cd76f4df903603f0da1ff6d2f917a)?

**Comment by charles-cooper** [comment]
note: review in group with https://github.com/vyperlang/titanoboa/pull/174

File: boa/interpret.py [hunk 3]
```diff
@@ -37,6 +38,15 @@ def set_cache_dir(cache_dir="~/.cache/titanoboa"):
     _disk_cache = DiskCache(cache_dir, compiler_version)
 
 
+def create_deployer(*args, **kwargs):
```
Comment (charles-cooper) [inline boa/interpret.py:41]: yea hm, this is going to blow up as soon as the user uses both zksync env and native env in the same program

File: boa/contracts/vyper/compiler_utils.py [hunk 4]
```diff
@@ -104,43 +109,52 @@ def generate_bytecode_for_internal_fn(fn):
...
         @external
         @payable
         def __boa_debug__() {return_sig}:
             {debug_body}
     """
     )
-    return compile_vyper_function(wrapper_code, contract)
+
+
+def detect_expr_type(source_code, contract):
```
Comment (charles-cooper) [inline boa/contracts/vyper/compiler_utils.py:150]: i'm not really sure why this needs to be refactored out

Comment (DanielSchiavini) [reply boa/contracts/vyper/compiler_utils.py:150]: We use this function directly for zksync eval

File: boa/interpret.py (hunk 3 above)
Comment (DanielSchiavini) [reply boa/interpret.py:41]: Well people can always create deployers "manually"
Also it should be fine to load contracts and switch deployers in between

File: boa/contracts/vyper/compiler_utils.py (hunk 4 above)
Comment (DanielSchiavini) [reply boa/contracts/vyper/compiler_utils.py:150]: 🤣 not sure.. thats for another review tho
This refactor is just separating concerns

Comment (charles-cooper) [reply boa/contracts/vyper/compiler_utils.py:150]: i don't think we should separate concerns until the need for it is demonstrated. if we aren't decided on whether the zksync plugin should have it, maybe the best solution for now is to implement a separate copy of `detect_expr_type` within the zksync plugin

Comment (DanielSchiavini) [reply boa/contracts/vyper/compiler_utils.py:150]: I still like the refactor but the code isn't too long, i can copy it

File: boa/interpret.py (hunk 3 above)
Comment (DanielSchiavini) [reply boa/interpret.py]: This method specifically isn't needed. I removed it and added more comments to the set_deployer_class

File: boa/contracts/abi/abi_contract.py [hunk 5]
```diff
@@ -34,7 +33,8 @@ def __init__(self, abi: dict, contract_name: str):
 
     @property
     def name(self) -> str:
-        return self._abi["name"]
+        # note: the `constructor` definition does not have a name
+        return self._abi.get("name") or self._abi["type"]
```
Comment (charles-cooper) [inline boa/contracts/abi/abi_contract.py:37]: what happens when we return `"constructor"` here? does anything rely on the value of name, or is it just for printing?

Comment (charles-cooper) [reply boa/contracts/abi/abi_contract.py:37]: probably returning `self._abi.get("name")` is more robust, caller has to handle the `None` case explicitly

File: boa/contracts/abi/abi_contract.py [hunk 6]
```diff
@@ -274,13 +281,17 @@ def stack_trace(self, computation: ComputationAPI) -> StackTrace:
         """
         Create a stack trace for a failed contract call.
         """
+        reason = ""
+        if computation.is_error and any(computation.error.args):
```
Comment (charles-cooper) [inline boa/contracts/abi/abi_contract.py:285]: what does `any(computation.error.args)` do?

File: boa/contracts/vyper/vyper_contract.py [hunk 7]
```diff
@@ -303,6 +305,9 @@ def _check(cond, msg=""):
         assert len(args) == 1, "multiple args!"
         assert len(kwargs) == 0, "can't mix args and kwargs!"
         err = args[0]
+        if isinstance(frame, str):
+            # frame for unknown contracts is a string
+            return _check(err in frame, f"{frame} does not match {args}")
```
Comment (charles-cooper) [inline boa/contracts/vyper/vyper_contract.py:310]: `_check` never returns anything

File: boa/contracts/vyper/vyper_contract.py [hunk 8]
```diff
@@ -315,6 +320,10 @@ def _check(cond, msg=""):
     # try to match a specific kwarg
     assert len(kwargs) == 1 and len(args) == 0
 
+    if isinstance(frame, str):
```
Comment (charles-cooper) [inline boa/contracts/vyper/vyper_contract.py:323]: why do we need this new check?

File: boa/integrations/jupyter/browser.py [hunk 9]
```diff
@@ -224,9 +211,31 @@ def _parse_js_result(result: dict) -> Any:
     if "data" in result:
         return result["data"]
 
+    def _find_key(input_dict, target_key, typ) -> Any:
+        for key, value in input_dict.items():
+            if isinstance(value, dict):
+                found = _find_key(value, target_key, typ)
+                if found is not None:
+                    return found
+            if key == target_key and isinstance(value, typ) and value != "error":
```
Comment (charles-cooper) [inline boa/integrations/jupyter/browser.py:220]: what happens when you call `_find_key(error, "error", str)` a few lines below?

File: boa/interpret.py [hunk 10]
```diff
@@ -86,22 +91,25 @@ def _ifaces():
...
-        ret = CompilerData(source_code, contract_name, interface_codes=ifaces, **kwargs)
+        ret = deployer.create_compiler_data(
+            source_code, contract_name, interface_codes=ifaces, **kwargs
+        )
         with anchor_compiler_settings(ret):
             _ = ret.bytecode, ret.bytecode_runtime  # force compilation to happen
         return ret
 
-    cache_key = str((contract_name, source_code, kwargs))
+    cache_key = str((contract_name, source_code, kwargs, deployer))
```
Comment (charles-cooper) [inline boa/interpret.py:107]: do we know that `deployer` has a sane repr? it could be any callable. we should at least have it on a separate line to be a bit clearer
```suggestion
    assert isinstance(deployer, type)
    deployer_id = repr(deployer)  # arbitrarily choose a unique str for the deployer
    cache_key = str((contract_name, source_code, kwargs, deployer_id))
```

File: boa/network.py [hunk 11]
```diff
@@ -43,8 +43,7 @@ def returndata_bytes(self):
     def is_error(self):
         if "structLogs" in self.raw_trace:
             return self.raw_trace["failed"]
-        else:
-            return "error" in self.raw_trace
+        return self.raw_trace.get("error") is not None
```
Comment (charles-cooper) [inline boa/network.py:46]: looks like a semantics-changing change. why is this necessary?

Comment (charles-cooper) [inline boa/network.py:46]: looks like a semantics-changing change. why is this necessary?

File: tests/unitary/jupyter/test_browser.py [hunk 12]
```diff
@@ -205,8 +206,46 @@ def test_browser_rpc_server_error(
...
+
+
+def test_browser_rpc_debug_error(mock_callback, env):
+    message = 'The method "debug_traceCall" does not exist / is not available.'
+    error = {
+        "error": {
+            "code": -32601,
+            "message": message,
+            "data": {
+                "origin": "https://44wgpcbsrwx-496ff2e9c6d22116-0-colab.googleusercontent.com",
```
Comment (charles-cooper) [inline tests/unitary/jupyter/test_browser.py:236]: a comment how this data was generated would be helpful

File: tests/unitary/jupyter/test_browser.py [hunk 13]
```diff
@@ -205,8 +206,46 @@ def test_browser_rpc_server_error(
     assert str(exc_info.value) == "-32603: server error"
 
 
+def test_browser_rpc_internal_error(mock_callback, env):
+    error = {
+        "code": -32603,
+        "message": "Internal JSON-RPC error.",
+        "data": {
+            "code": 3,
+            "message": "insufficient funds for gas + value. balance: 0, fee: 116, value: 0",
```
Comment (charles-cooper) [inline tests/unitary/jupyter/test_browser.py:215]: a comment how this data was generated would be helpful. in both of these cases i think it would be good to add a comment saying we are agnostic to the exact data, we are just testing that we find the key properly.

File: boa/integrations/jupyter/browser.py (hunk 9 above)
Comment (charles-cooper) [reply boa/integrations/jupyter/browser.py:220]: ah the check is for `value == "error"`, not `key == "error"`. what is the purpose of the check then?

Comment (DanielSchiavini) [reply boa/integrations/jupyter/browser.py:220]: The RPCs return inconsistent errors. This code is trying to find the best error to describe the situation.
For example, this test returns `error.value="error"`:
https://github.com/vyperlang/titanoboa/blob/2dfdf53fff6584504c8ecb6ce7712d3062aa1d3d/tests/unitary/jupyter/test_browser.py#L189

File: boa/network.py (hunk 11 above)
Comment (DanielSchiavini) [reply boa/network.py:46]: zksync always returns `"error=None"`

Comment (DanielSchiavini) [reply boa/network.py:46]: duplicated comment

Comment (charles-cooper) [reply boa/network.py:46]: spooky. in that case, comment would be good
```suggestion
        # we can have `"error": null` in the payload
        return self.raw_trace.get("error") is not None
```

File: boa/interpret.py (hunk 10 above)
Comment (charles-cooper) [reply boa/interpret.py:107]: if it's a function, the repr will have a pointer. it is more likely to produce a false cache miss than a false cache hit, but it's still possible.

File: boa/contracts/abi/abi_contract.py [hunk 14]
```diff
@@ -62,6 +62,8 @@ def pretty_signature(self) -> str:
 
     @cached_property
     def method_id(self) -> bytes:
+        if self._abi["type"] == "constructor":
+            return b""  # constructors don't have method IDs
```
Comment (charles-cooper) [inline boa/contracts/abi/abi_contract.py:66]: i think this should be handled at the caller really

File: boa/contracts/abi/abi_contract.py [hunk 15]
```diff
@@ -34,7 +33,8 @@ def __init__(self, abi: dict, contract_name: str):
 
     @property
     def name(self) -> str:
-        return self._abi["name"]
+        # note: the `constructor` definition does not have a name
+        return self._abi.get("name", "")
```
Comment (charles-cooper) [inline boa/contracts/abi/abi_contract.py:37]: yea i don't think we should use `""` as a default here, i think we should return Optional and just handle the None case at the call sites. like we will get stuff like `""` as a key in the overloads dictionary, which seems funky.

or if we really want a name, we can use `"__init__"` (or `"constructor"` as i think you originally had).

Comment (DanielSchiavini) [reply boa/contracts/abi/abi_contract.py:37]: yeah constructor was the first workaround I thought of, because it's just clear and simple.
handling it at the call sites is fine, but it creates a lot of boiler plate code

**Review by charles-cooper — APPROVED** [review]
thank you !!!!
